core/engine/src/object/jsobject.rs

```diff
@@ -92,9 +119,11 @@ impl JsObject {
             vtable,
         });
 
-        Self {
-            inner: coerce_gc(gc),
-        }
+        // SAFETY: The pointer is guaranteed to be valid because we just created it.
+        // `VTableObject<ErasedObjectData>` and `VTableObject<T>` have the same size and alignment.
+        let inner = unsafe { Gc::cast_unchecked::<ErasedVTableObject>(gc) };
+
+        JsObject { inner }
```

Comment:
```suggestion
        JsObject { inner: gc }.upcast()
```

We could eliminate the unsafe by using the `upcast` method which already has the unsafe code.